cupy/elementwise.py

```diff
@@ -501,19 +508,43 @@ def _guess_routine_from_dtype(ops, dtype):
 
 def _guess_routine(name, cache, ops, in_args, dtype):
     if dtype is None:
-        key = tuple([numpy.dtype(type(i)).type
-                     if isinstance(i, _python_scalar_type) else i.dtype.type
-                     for i in in_args])
-    else:
-        key = dtype
-
-    op = cache.get(key, ())
-    if op is ():
-        if dtype is None:
-            op = _guess_routine_from_in_types(ops, key)
+        all_scalars = True
+        max_array_kind = -1
+        max_scalar_kind = -1
+        for i in in_args:
+            if isinstance(i, cupy.ndarray):
+                kind = _kind_score[i.dtype.kind]
+                all_scalars = False
+                if kind > max_array_kind:
+                    max_array_kind = kind
+            else:
+                if isinstance(i, _python_scalar_type):
+                    dtype = numpy.dtype(type(i))
+                else:
+                    dtype = i.dtype
+                kind = _kind_score[dtype.kind]
+                if kind > max_scalar_kind:
+                    max_scalar_kind = kind
+        use_min_scalar = not all_scalars and max_array_kind >= max_scalar_kind
```

Comment:
Please extract the routine for determining the value of `use_min_scalar` as a separate function.